Mathlib/Geometry/Manifold/Sheaf/Smooth.lean

```diff
@@ -0,0 +1,403 @@
+  sheaf of smooth functions into a [Lie group]/[abelian Lie group]/[smooth ring]/[smooth commutative
+  ring], as a sheaf of [groups]/[abelian groups]/[rings]/[commutative rings]
+* `smoothSheafCommRing.forgetStalk`: Identify the stalk at a point of the sheaf-of-commutative-rings
+  of functions from `M` to `R` (for `R` a smooth ring) with the stalk at that point of the
+  corresponding sheaf of types.
+* `smoothSheafCommRing.eval`: upgrade `smoothSheaf.eval` to a ring homomorphism when considering the
+  sheaf of smooth functions into a smooth commutative ring
+* `smoothSheafCommGroup.compLeft`: For a manifold `M` and a smooth homomorphism `φ` between
+  abelian Lie groups `A`, `A'`, the 'postcomposition-by-`φ`' morphism of sheaves from
+  `smoothSheafCommGroup IM I M A` to `smoothSheafCommGroup IM I' M A'`
+
+# Main results
+
+* `smoothSheaf.eval_surjective`: `smoothSheaf.eval` is surjective.
+* `smoothSheafCommRing.eval_surjective`: `smoothSheafCommRing.eval` is surjective.
+
+## TODO
+
+There are variants of `smoothSheafCommGroup.compLeft` for `GroupCat`, `RingCat`, `CommRingCat`;
+this is just boilerplate and can be added as needed.
+
+Similarly, there are variants of `smoothSheafCommRing.forgetStalk` and `smoothSheafCommRing.eval`
+for `GroupCat`, `CommGroupCat` and `RingCat` which can be added as needed.
+
+Currently there is a universe restriction: one can consider the sheaf of smooth functions from `M`
+to `N` only if `M` and `N` are in the same universe.  For example, since `ℂ` is in `Type`, we can
+only consider the structure sheaf of complex manifolds in `Type`, which is unsatisfactory. The
+obstacle here is in the underlying category theory constructions, and there is WIP (as of June 2023)
+to fix this.  See
+https://github.com/leanprover-community/mathlib/pull/19153
+and cross-references there.
+-/
+
+
+noncomputable section
+open TopologicalSpace Opposite
+
+universe u
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {EM : Type*} [NormedAddCommGroup EM] [NormedSpace 𝕜 EM]
+  {HM : Type*} [TopologicalSpace HM] (IM : ModelWithCorners 𝕜 EM HM)
+  variable {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {H : Type*} [TopologicalSpace H] (I : ModelWithCorners 𝕜 E H)
+  {H' : Type*} [TopologicalSpace H'] (I' : ModelWithCorners 𝕜 E H')
+  (M : Type u) [TopologicalSpace M] [ChartedSpace HM M]
+  (N G A A' R : Type u) [TopologicalSpace N] [ChartedSpace H N]
+  [TopologicalSpace G] [ChartedSpace H G] [TopologicalSpace A] [ChartedSpace H A]
+  [TopologicalSpace A'] [ChartedSpace H' A'] [TopologicalSpace R] [ChartedSpace H R]
+
+section TypeCat
+
+/-- The sheaf of smooth functions from `M` to `N`, as a sheaf of types. -/
+def smoothSheaf : TopCat.Sheaf (Type u) (TopCat.of M) :=
+  (contDiffWithinAt_localInvariantProp IM I ⊤).sheaf M N
+
+variable {M}
+
+instance smoothSheaf.coeFun (U : (Opens (TopCat.of M))ᵒᵖ) :
+    CoeFun ((smoothSheaf IM I M N).val.obj U) (fun _ ↦ ↑(unop U) → N) :=
```

Comment:
Is this a job for `TopCat.Sheaf.presheaf`:
```suggestion
    CoeFun ((smoothSheaf IM I M N).presheaf.obj U) (fun _ ↦ ↑(unop U) → N) :=
```
(and also in 21 more occurrences below)

Reply:
Done!